### PR TITLE
FIX Improving support for cascading themes

### LINKED ISSUE #392: Cascading Themes?
Is it possible to use cascading themes with subsites? Or to define the allowed themes from yml? It looks like they might currently be able to be set from config.php, but it would be super handy if we could define a list of themes from yml. eg:

```
SilverStripe\Subsites\Model\Subsite:
  allowed_themes:
    - '/theme-default'
    - '/theme-other'
```

Or, if cascading needs to be defined: 
```
SilverStripe\Subsites\Model\Subsite:
  allowed_themes:
    defaultThemeName:
       - '$public'
       - '/theme-default'
       - '$default'
    otherThemeName:
       - '$public'
       - '/theme-other'
       - '$default'
```

### PR DESCRIPTION
**Pull request by ScopeyNZ**
- Fixes an issue where themes would cascade "up" the list of themes
- Provides configuration for defining custom theme options with their own sets of cascading themes

Fixes #392

### CODE REVIEW

File: src/Service/ThemeResolver.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     *   'theme-1' => [
+     *     '$public',
+     *     'starter',
+     *     '$default',
+     *   ],
+     *   'theme-2' => [
+     *     'custom',
+     *     'watea',
+     *     'starter',
+     *     '$public',
```
Comment (robbieaverill) [inline src/Service/ThemeResolver.php]: The public entry would always be at the start

Comment (ScopeyNZ) [reply src/Service/ThemeResolver.php]: Eh? You can choose.

File: tests/php/Service/ThemeResolverTest.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    public function testSubsiteWithCustomThemeDoesNotCascadeUpTheList()
+    {
+        $subsite = new Subsite();
+        $subsite->Theme = 'main';
+
+        $resolver = new ThemeResolver();
+
+        $expected = [
+            'main', // 'main' is moved to the top
+            SSViewer::PUBLIC_THEME, // $public is preserved
```
Comment (robbieaverill) [inline tests/php/Service/ThemeResolverTest.php]: I'm not sure this is a realistic use case - public is put first in theme lists

Comment (ScopeyNZ) [reply tests/php/Service/ThemeResolverTest.php]: This is how it works currently - the chosen theme is moved _above_ public. It'd technically be breaking if I changed this.

File: src/Model/Subsite.php [hunk 3]
```diff
@@ -783,7 +783,7 @@ public function fieldLabels($includerelations = true)
      */
     public function allowedThemes()
     {
-        if ($themes = self::$allowed_themes) {
+        if (($themes = self::$allowed_themes) || ($themes = ThemeResolver::singleton()->getCustomThemeOptions())) {
```
Comment (robbieaverill) [inline src/Model/Subsite.php]: Do you want to make a separate PR to master to deprecate `$allowed_themes`?

Comment (ScopeyNZ) [reply src/Model/Subsite.php]: Yes - I was going to deprecate this and another useless extension point after I merge up to `2.3`.

**Comment by robbieaverill** [comment]
Merge on green

**Comment by robbieaverill** [comment]
Ah, the test failure is valid - the public theme constant was not available in 4.0

**Comment by ScopeyNZ** [comment]
Ok there's still stuff to fix with that `$public` stuff. I'll sort it out later. I think it's important that it's handled consistently (compared to previously) with the new code - but I'll have to figure out a way to incorporate that into tests that run with and without `$public` support.

**Comment by ScopeyNZ** [comment]
I cheated by rebasing onto `2.1`. Hopefully tests are green after the latest commit.

**Pull request merged**